android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt

```diff
@@ -499,42 +470,72 @@ open class FxaAccountManager(
     }
 
     /**
-     * Side-effects of entering [AccountState] type states - our stable states.
-     * Once we reach these states, we simply notify our observers.
+     * Side-effects of entering [AccountState] type states
+     *
+     * Upon entering these states, observers are typically notified. The sole exception occurs
+     * during the completion of authentication, where it is necessary to populate the
+     * SyncAuthInfoCache for the background synchronization worker. Should this process fail, the
+     * system will transition to `AccountState.AuthenticationProblem`.
      */
-    private suspend fun accountStateSideEffects(forState: State.Idle, via: Event): Unit = when (forState.accountState) {
+    private suspend fun accountStateSideEffects(
+        forState: State.Idle,
+        via: Event,
+    ): Event? = when (forState.accountState) {
         AccountState.NotAuthenticated -> when (via) {
             Event.Progress.LoggedOut -> {
+                resetAccount()
                 notifyObservers { onLoggedOut() }
+                null
             }
             Event.Progress.FailedToBeginAuth -> {
+                resetAccount()
                 notifyObservers { onFlowError(AuthFlowError.FailedToBeginAuth) }
+                null
             }
             Event.Progress.FailedToCompleteAuth -> {
+                resetAccount()
                 notifyObservers { onFlowError(AuthFlowError.FailedToCompleteAuth) }
+                null
             }
             Event.Progress.FailedToRecoverFromAuthenticationProblem -> {
                 finishedAuthRecovery()
+                null
             }
-            else -> Unit
+            else -> null
         }
         AccountState.Authenticated -> when (via) {
             is Event.Progress.CompletedAuthentication -> {
-                notifyObservers { onAuthenticated(account, via.authType) }
-                refreshProfile(ignoreCache = false)
-                Unit
+                val operation = when (via.authType) {
+                    AuthType.Existing -> "CompletingAuthentication:accountRestored"
+                    else -> "CompletingAuthentication:AuthData"
+                }
+                if (authenticationSideEffects(operation)) {
+                    notifyObservers { onAuthenticated(account, via.authType) }
+                    refreshProfile(ignoreCache = false)
+                    null
+                } else {
+                    Event.Account.AccessTokenKeyError
```

Comment:
I think those are still fired if `finalizeDevice()` fails and we would never get to the `Authenticated` state in that case.

I'm not sure if we discussed this, but I feel the opposite about that call compared to `getAccessToken`.   Clients are going to want to finalize their device as part of logging in, so it should happen inside the state machine.  The means that error handling will also happen inside the state machine rather than in `accountStateSideEffects()`.  WDYT?

Reply:
It sounds reasonable to me as well, but it's a change to the state flow – and I am not sure what are the consequences. Given that PR is focusing on decoupling, maybe we shouldn't be introducing any flow changes? From the top of my head, `State.Idle(AccountState.NotAuthenticated)` and `State.Idle(AccountState.AuthenticationProblem)` might have different UI representations. If we were to do the change, we should do it intentionally and test how it affects the UX.

Why not emitting the same event on `authenticationSideEffects`? What has changed? It looks like the function itself didn't change.

Yep, errors of "fatal" level is what we are hunting during health monitoring. Let's address it in a separate PR though. As a side note, I wonder if there should be some process to accompany the addition of a new exception that A-S might through, as those changes are very likely to crash the app.

Reply:
I think something needs to change because `authenticationSideEffects` is now called from the `AccountState.Authenticated` state rather than `ProgressState.CompletingAuthentication`.  That's why I was suggesting `Logout` as the event.   Alternatively, we could emit the same event and add a new transition.

Reply:
I just updated this to return the `Logout` event which seems better than `AccessTokenKeyError`, but I'm still open to other suggestions.